java-checks/src/main/java/org/sonar/java/checks/spring/BeanMethodOfNonProxiedSingletonInvocationCheck.java

```diff
@@ -0,0 +1,121 @@
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+
+@Rule(key = "S6838")
+public class BeanMethodOfNonProxiedSingletonInvocationCheck extends IssuableSubscriptionVisitor {
+  private static final String CONFIGURATION_ANNOTATION = "org.springframework.context.annotation.Configuration";
+  private static final String SCOPE_ANNOTATION = "org.springframework.context.annotation.Scope";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    boolean isTargetClass = getConfigurationAnnotation((ClassTree) tree)
+      .map(BeanMethodOfNonProxiedSingletonInvocationCheck::hasProxyBeanMethodsDisabled)
+      .orElse(Boolean.FALSE);
+    if (!isTargetClass) {
+      return;
+    }
+    var visitor = new NonProxiedMethodInvocationVisitor((ClassTree) tree);
+    tree.accept(visitor);
+    visitor.locations.forEach(invocation -> reportIssue(invocation, "Replace this bean method invocation " + "with a dependency injection."));
+  }
+
+  private static Optional<AnnotationTree> getConfigurationAnnotation(ClassTree tree) {
+    SymbolMetadata metadata = tree.symbol().metadata();
+    for (SymbolMetadata.AnnotationInstance instance : metadata.annotations()) {
+      if (instance.symbol().type().is(CONFIGURATION_ANNOTATION)) {
+        return Optional.ofNullable(metadata.findAnnotationTree(instance));
+      }
+    }
+    return Optional.empty();
+  }
+
+
+  private static boolean hasProxyBeanMethodsDisabled(AnnotationTree annotation) {
+    return annotation.arguments().stream()
+      .filter(argument -> argument.is(Tree.Kind.ASSIGNMENT))
+      .map(AssignmentExpressionTree.class::cast)
+      .anyMatch(BeanMethodOfNonProxiedSingletonInvocationCheck::setsProxyBeanMethodsToFalse);
+  }
+
+  private static boolean setsProxyBeanMethodsToFalse(AssignmentExpressionTree assignment) {
+    return "proxyBeanMethods".equals(((IdentifierTree) assignment.variable()).name()) &&
+      Boolean.FALSE.equals(ExpressionsHelper.getConstantValueAsBoolean(assignment.expression()).value());
+  }
+
+  static class NonProxiedMethodInvocationVisitor extends BaseTreeVisitor {
+    private final ClassTree parentClass;
+    private final List<MethodInvocationTree> locations = new ArrayList<>();
+
+    NonProxiedMethodInvocationVisitor(ClassTree parentClass) {
+      this.parentClass = parentClass;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      super.visitMethodInvocation(tree);
+      MethodTree declaration = tree.methodSymbol().declaration();
+      if (declaration == null) {
+        return;
+      }
+      if (returnsAPrototypeBean(declaration)) {
+        return;
+      }
```

Comment:
Maybe combine this into one condition?

```suggestion
      if (declaration == null || returnsAPrototypeBean(declaration)) {
        return;
      }
```